### ISSUE TITLE
Gateway support for /ipfs/{cid}?format=car|block|...

### ISSUE THREAD

**Issue by lidel** [ipfs/go-ipfs#8234]
>  @mikeal @olizilla @autonome  @warpfork @aschmahmann – this is a quick memo with my initial synthesis of the `?format=`  idea. Bit thin on details, but want to get early feedback/temperature check  before I start elaborating on this in gateway specs in the following weeks.

This is a meta issue for streamlining various feature requests and needs under a single opt-in query parameter that enables gateway users to fetch a specific representation of a specific content path.

Support for each format can be discussed/added via a separate issue/PR – this issue is just for tracking the bigger picture around unified `format` paramerer.

# MVP formats

## Ability to fetch every CID as full DAG in CAR or a single Block

This is the key  feature to enable [Verifiable Gateway Responses](https://github.com/ipfs/in-web-browsers/issues/128)  and "HTTP-based transport for IPFS"  (mobile browsers, IoT) without introducing even more dependency on `/api/v0`, and giving us flexibility for adding new features in the future.

- [ ] `?format=car`
  - Returns binary  `application/octet-stream` with CAR archive for entire DAG behind the content path
  - Supersedes  `/api/v0/dag/export`, but with better UX: 
    - works on DNSLink websites that do not expose `/api/v0`
    - `content-disposition` defaults to  `{filename|cid}.car` 
- [ ] `?format=block` 
  - Returns binary `application/octet-stream` with the root block identified by CID
  - Supersedes  `/api/v0/block/get`, but with better UX:
    -  works on DNSLink websites that do not expose `/api/v0`
    - `content-disposition` defaults to  `{cid}.block` 

## CBOR / JSON

The idea here is to unlock the power of IPLD by making non-dag-pb codecs available over gateway with built-in bi-directional conversion: a CID that is `dag-cbor` or `dag-json` could be fetched as either.

We  most likely want to do this in https://github.com/ipfs/go-ipfs/pull/8037 (refactor to use `?format=`)

- [ ] `?format=dag-cbor` 
  - Returns `dag-cbor` binary data as `application/cbor`
- [ ] `?format=dag-json`
  - Returns `dag-json` document as  `application/json`
  - This is especially useful for getting `dag-json` representation of `dag-cbor` document for use in web apps

Sidenote: if we find a way to  represent `dag-pb`   directory as `dag-cbor` document, then `?format=dag-json` could be an elegant way for supporting [JSON responses for directory listing](https://github.com/ipfs/go-ipfs/issues/7552), which was also requested by our users. 

## Future ideas / lower priorities

- `?format=tar`
  -  https://github.com/ipfs/go-ipfs/issues/7746 
  -  https://github.com/ipfs/go-ipfs/issues/7959
- `?format=share-img`
  -  https://github.com/ipfs/dir-index-html/issues/51
  
## Behaviors

- [ ] `?format` missing 
    -  if codec is `dag-pb` or `raw` return file/directory 
       (current gateway behavior)
    - else (codec without default behavior), return error suggesting passing `?format=car|block|..`

----

**Comment by mikeal** [comment]
I like this querystring approach a lot. You can easily imagine extending this with new parameters for partial dag queries and the like.

One thing I’d like to specify is how an implementation exposes what formats it does and does not support. Then clients can implement fallback logic in order to be more robust and servers aren’t required to implement every feature ever.

**Comment by olizilla** [comment]
It looks promising! Does it matter that it'd be mixing codecs `?format=dag-json` and containers `?format=car`. How do we deal with verisoning, would CARv2 be `?format=carv2`? It's not called out in the issue, but is the plan to also honour an  `Accept` header if provided.

> else (codec without default behavior), return error suggesting passing ?format=car|block|..

can a `dag-json` be returned as json without needing to specify the format? I know this has come up before, but I can't recall why that would be bad.

and can has mime/types https://github.com/ipld/specs/issues/368

**Comment by ribasushi** [comment]
> but I can't recall why that would be bad.

@olizilla relevant 🧵 
https://github.com/ipfs/go-ipfs/pull/8037#discussion_r604090502

**Comment by lidel** [comment]
- **On default behavior** 
  - `car` and `block` would work for every CID, other things like json or cbor will be available per-codec-basis
  - When requested CID is not available in specified format (or does not have a default one), a  human-readable error message suggesting appending explicit `?format=dag-json|dag-dbor|block|car` is returned. 
  - To facilitate things like automated fallback or format discovery via HEAD request, the list of supported formats could be included as HTTP header such as  `Link` from https://github.com/ipfs/in-web-browsers/issues/179:
    ```
    Link: <ipfs://bafy?format=block>; rel=describedby; type="application/octet-stream"
    Link: <ipfs://bafy?format=car>; rel=describedby; type="application/octet-stream"    
    Link: <ipfs://bafy?format=dag-json>; rel=describedby; type="application/json"
    Link: <ipfs://bafy?format=dag-cbor>; rel=describedby; type="application/cbor"
    ```
  - The concern about default behavior was about binary `dag-cbor`  (thread linked above by @ribasushi). 
    - For  `dag-cbor` the default state will be error, as there is no obvious default response, and we may want to render some GUI on gateways in the future.
    - Returning `application/json`  for  `dag-json`  by default (without explicit `?format`) is ok.
- **On mixing codecs and containers:**  my reasoning for a single `?format=` is that both codecs and containers translate to distinct _response formats_. HTTP client does not care about IPFS-specific taxonomy, it requests specific thing in specific format (`Accept` or `?format=` ) and gets it. 
  - Sidenote: I used explicit `dag-cbor` and `dag-json` just to highlight that CIDs will be traversable thanks to IPLD conventions, but we may shorten this to `json` and `cbor` to improve UX.
- **On CAR versions:** AFAIK we have built-in versioning in CARs: CARv2 will includes version in the header in a backward-compatible way, so CARv1 parser will return "unsupported version" for CARv2.  Due to this I see no need for versioning  here.

**Comment by lidel** [comment]
Food for thought: I felt mixing IPLD codec names with formats like  car and block bit confusing.
Perhaps we should make IPLD codec override (because one is already in the CID) more explicit:

```
?format=car 
?format=block

?format=ipld
?format=ipld&codec=dag-json
?format=ipld&codec=dag-cbor
```

**Comment by mikeal** [comment]
what ever happened to this?

**Comment by lidel** [comment]
Lack of bandwidth within stewards group – it is still on the roadmap

**Comment by lidel** [comment]
Note:  `car` code was added  in https://github.com/multiformats/multicodec/pull/258 and the discussion around its meanign and purpose continues in https://github.com/multiformats/multicodec/issues/239#issuecomment-1046124811

My take / question: can we use the codec field of CIDv1 to indicate expected format/transformation when requesting data from Grateways?

> [..]  convention where `raw` and  `car` codecs are used on HTTP  Gateway **as a way of requesting a single Block or a CAR with blocks for a DAG**.
> - HTTP `GET /ipfs/{cid-with-raw-codec}` returning a raw Block
> - HTTP `GET /ipfs/{cid-with-car-codec}` returning a CAR with the entire DAG behind a CID 
> 
> In this convention the multihash in a CID represents the root block of a DAG, and if you plan to use `car` [code] with a multihash that has different meaning, we should agree on that now.

We could play it safe and use `?format=car` (or shorter `?as=car`)  for now, 
but things may be more intuitive if `?format=car` returns a redirect to CIDv1 with `car` codec, and that would return a CAR stream.

**Comment by lidel** [comment]
Block/CAR response types are implemented in https://github.com/ipfs/go-ipfs/pull/8758 – ready for review, plan is to ship it in go-ipfs 0.13.
